### ISSUE TITLE
求助

### ISSUE THREAD

**Issue by XinDong10** [joellliu/SegmentAndComplete#8]
老师，还想再求助一次，我可以问问apricotmask数据集中dev集合里面每个数据的类别‘-10’是代表哪一类嘛。

**Comment by joellliu** [comment]
应该是代表patch，可以把bounding box在图上画出来确认一下。

**Comment by XinDong10** [comment]
老师，确实是表示patch，但是我如何知道这个patch的target类别是哪一类呢，看您计算了target成功率需要这个类别信息

**Comment by joellliu** [comment]
这个需要在armory里面通过查询patch_id得到
https://github.com/twosixlabs/armory/blob/4a5e808bfa02952618da9ca7d5ae06f793775a11/armory/data/adversarial/apricot_metadata.py#L28

**Comment by joellliu** [comment]
evaluation函数可以参考这个
https://github.com/twosixlabs/armory/blob/4a5e808bfa02952618da9ca7d5ae06f793775a11/armory/metrics/task.py#L1920

**Comment by alex1243423** [comment]
老师您好，可以问下针对apricot训练的patch检测模型，可以方便公开嘛，我的邮箱是2823627395@qq.com。非常期待您的回复。

**Comment by joellliu** [comment]
> 老师您好，可以问下针对apricot训练的patch检测模型，可以方便公开嘛，我的邮箱是[2823627395@qq.com](mailto:2823627395@qq.com)。非常期待您的回复。

您好！可以的，我过阵子上传一下。

**Comment by XinDong10** [comment]
谢谢老师，您太优秀了，可以问下您dpatch使用的是armory库的接口嘛，我这里生成的dpatch并没有使MAP下降。老师还想问下dpatch是都在左上角嘛

**Comment by joellliu** [comment]
> 谢谢老师，您太优秀了，可以问下您dpatch使用的是armory库的接口嘛，我这里生成的dpatch并没有使MAP下降。老师还想问下dpatch是都在左上角嘛

dpatch用的这个代码 https://github.com/joellliu/SegmentAndComplete/blob/main/dpatch.py
dpatch的位置和pgd一样，是随机生成并存在数据集里了，这样可以保证不同攻击方法都在同一位置。

**Comment by joellliu** [comment]
> 老师您好，可以问下针对apricot训练的patch检测模型，可以方便公开嘛，我的邮箱是[2823627395@qq.com](mailto:2823627395@qq.com)。非常期待您的回复。

您好！模型已经上传了 https://github.com/joellliu/SegmentAndComplete/tree/main#model-checkpoints

**Comment by XinDong10** [comment]
谢谢您老师，我今天尝试了一下，可以问下您也是只使用generate代码，没有使用apply_patch嘛，是这样的嘛。
frcnn = PyTorchFasterRCNN(
    # 3 channels (RGB) and size 640x640
    input_shape=(3, 640, 640),
    # Pixel values should be clipped between 0 and 255
    clip_values=(0, 255),
    # Color channels are the last dimension
    channels_first=False,
    # Use all the losses provided by Faster R-CNN model
    attack_losses=("loss_classifier", "loss_box_reg", "loss_objectness", "loss_rpn_box_reg"),
    # Use the GPU for computation
    device_type="gpu"
)

# Create the RobustDPatch attack
attack = DPatch1(
    # Our trained object detector, here Faster R-CNN model
    estimator=frcnn,
    # (height, width, nb_channels)
    patch_shape=(75, 75, 3),
    # Number of optimization steps
    learning_rate = 0.01,
    max_iter=200
)
data  = torch.load("75/"+"data_"+name+".pth.tar")
x = data['x'][0].permute(1, 2, 0).unsqueeze(0).numpy() #HWC RGB
#print(x.shape)
mx = data['y'][0]
mask = data['y'][0]['masks']
xmin, ymin = data['idx'][0, :2]
xmin = int(xmin)
ymin = int(ymin)
patched_image = attack.generate(x,y=mx,xmin=xmin,ymin=ymin)
adversarial_image = Image.fromarray(np.squeeze((patched_image*255).astype(np.uint8))
老师老师，我还发现dpatch.py里面的“patch_gradients_i = gradients[i_image, i_x_1:i_x_2，i_y_1:i_y_2, :]”这里或者使用i_y_1这些变量的地方，是不是该写成“patch_gradients_i = gradients[i_image, i_y_1:i_y_2， i_x_1:i_x_2，:]”呢？因为我看gradients的大小是1*h*w*3的。所以我就改了，生成了一次，效果不好，可能是我改了的原因

**Comment by joellliu** [comment]
> 谢谢您老师，我今天尝试了一下，可以问下您也是只使用generate代码，没有使用apply_patch嘛，是这样的嘛。 frcnn = PyTorchFasterRCNN( # 3 channels (RGB) and size 640x640 input_shape=(3, 640, 640), # Pixel values should be clipped between 0 and 255 clip_values=(0, 255), # Color channels are the last dimension channels_first=False, # Use all the losses provided by Faster R-CNN model attack_losses=("loss_classifier", "loss_box_reg", "loss_objectness", "loss_rpn_box_reg"), # Use the GPU for computation device_type="gpu" )
> 
> # Create the RobustDPatch attack
> attack = DPatch1( # Our trained object detector, here Faster R-CNN model estimator=frcnn, # (height, width, nb_channels) patch_shape=(75, 75, 3), # Number of optimization steps learning_rate = 0.01, max_iter=200 ) data = torch.load("75/"+"data_"+name+".pth.tar") x = data['x'][0].permute(1, 2, 0).unsqueeze(0).numpy() #HWC RGB #print(x.shape) mx = data['y'][0] mask = data['y'][0]['masks'] xmin, ymin = data['idx'][0, :2] xmin = int(xmin) ymin = int(ymin) patched_image = attack.generate(x,y=mx,xmin=xmin,ymin=ymin) adversarial_image = Image.fromarray(np.squeeze((patched_image_255).astype(np.uint8)) 老师老师，我还发现dpatch.py里面的“patch_gradients_i = gradients[i_image, i_x_1:i_x_2，i_y_1:i_y_2, :]”这里或者使用i_y_1这些变量的地方，是不是该写成“patch_gradients_i = gradients[i_image, i_y_1:i_y_2， i_x_1:i_x_2，:]”呢？因为我看gradients的大小是1_h_w_3的。所以我就改了，生成了一次，效果不好，可能是我改了的原因

是的，只用generate，不用apply_patch。H/W的顺序应该不会太影响performance。你测mAP的时候是用adversarial_image = Image.fromarray(np.squeeze((patched_image255).astype(np.uint8)) 这个测吗？这个会有量化损失可能会影响attack效果。
